shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,776 @@
+       * the type parameter `tp` in `target`.
+       */
+      bindingset[a, target]
+      pragma[inline_late]
+      private Type getTypeArgument(Access a, Declaration target, TypeParameter tp, TypePath path) {
+        exists(TypeArgumentPosition tapos, TypeParameterPosition tppos |
+          result = a.getTypeArgument(tapos, path) and
+          tp = target.getTypeParameter(tppos) and
+          typeArgumentParameterPositionMatch(tapos, tppos)
+        )
+      }
+
+      /**
+       * Holds if the type `t` at `path` of `a` matches the type parameter `tp`
+       * of `target`.
+       */
+      pragma[nomagic]
+      private predicate directTypeMatch(
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        exists(AccessPosition apos, DeclarationPosition dpos, TypePath pathToTypeParam |
+          adjustedAccessType(a, apos, target, pathToTypeParam.append(path), t) and
+          tp = target.getDeclaredType(dpos, pathToTypeParam) and
+          not exists(getTypeArgument(a, target, tp, _)) and
+          accessDeclarationPositionMatch(apos, dpos)
+        )
+      }
+
+      private module AccessBaseType {
+        private predicate relevantAccess(Access a, AccessPosition apos) {
+          exists(Declaration target |
+            adjustedAccessType(a, apos, target, _, _) and
+            target.getDeclaredType(_, _) instanceof TypeParameter
+          )
+        }
+
+        pragma[nomagic]
+        private Type inferRootType(Access a, AccessPosition apos) {
+          relevantAccess(a, apos) and
+          result = a.getInferredType(apos, TypePath::nil())
+        }
+
+        pragma[nomagic]
+        private Type inferTypeAt(Access a, AccessPosition apos, TypeParameter tp, TypePath suffix) {
+          relevantAccess(a, apos) and
+          exists(TypePath path0 |
+            result = a.getInferredType(apos, path0) and
+            path0.isCons(tp, suffix)
+          )
+        }
+
+        /**
+         * Holds if `baseMention` is a (transitive) base type mention of the type of
+         * `a` at position `apos`, and `t` is mentioned (implicitly) at `path` inside
+         * `base`. For example, in
+         *
+         * ```csharp
+         * class C<T1> { }
+         *
+         * class Base<T2> { }
+         *
+         * class Mid<T3> : Base<C<T3>> { }
+         *
+         * class Sub<T4> : Mid<C<T4>> { }
+         *
+         * new Sub<int>().ToString();
+         * ```
+         *
+         * for the node `new Sub<int>()`, which is the receiver of a method call, we
+         * have:
+         *
+         * `baseMention` | `path`    | `t`
+         * ------------- | --------- | ---
+         * `Mid<C<T4>>`  | `"0"`     | ``C`1``
+         * `Mid<C<T4>>`  | `"0.1"`   | `int`
+         * `Base<C<T3>>` | `"0"`     | ``C`1``
+         * `Base<C<T3>>` | `"0.0"`   | ``C`1``
+         * `Base<C<T3>>` | `"0.0.1"` | `int`
```

Comment:
Are these `1`s correct? Given that the examples so far have used positional, zero-based indices, shouldn't these `1`s be `0`s instead?

Reply:
Yes 🤦